File: sdks/java/reactor-sdk/src/main/java/inc/reactor/sdk/VideoFrame.java
package inc.reactor.sdk;

import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.util.Optional;
import org.jspecify.annotations.Nullable;

/**
 * One decoded video frame.
 *
 * <p><b>Valid only for the duration of the handler call.</b> The pixels belong to the FFI, which
 * reuses them the moment the handler returns. Keeping a frame and reading it later would be a
 * use-after-free in any other binding; here the underlying memory is scoped to the callback, so
 * doing it throws {@link IllegalStateException} instead. Copy what you keep — {@link
 * #toByteArray()} is the copy.
 *
 * <p>Handlers run inline on the FFI's delivery thread, on purpose. Blocking there is the
 * backpressure: the FFI keeps only the newest frame while the handler runs, so a slow handler drops
 * frames rather than growing a queue.
 */
public final class VideoFrame {

    private final MemorySegment pixels;
    private final int width;
    private final int height;
    private final long frameId;
    private final long timestampUs;
    private final byte @Nullable [] userData;

    VideoFrame(
            MemorySegment pixels, int width, int height, long frameId, long timestampUs, byte @Nullable [] userData) {
        this.pixels = pixels;
        this.width = width;
        this.height = height;
        this.frameId = frameId;
        this.timestampUs = timestampUs;
        this.userData = userData;
    }

    /** @return frame width in pixels */
    public int width() {
        return width;
    }

    /** @return frame height in pixels */
    public int height() {
        return height;
    }

    /**
     * The raw BGRA pixels — {@code width * height * 4} bytes, in B, G, R, A order.
     *
     * <p>Reading this after the handler has returned throws {@link IllegalStateException}.
     *
     * @return a read-only view of the frame's memory
     */
    public MemorySegment pixels() {
        return pixels;
    }

    /**
     * Copies the pixels onto the Java heap.
     *
     * <p>{@code width * height * 4} bytes — 8 MB for 1080p, so this is a real cost at frame rate.
     * Read {@link #pixels()} directly when the handler can do its work before returning.
     *
     * @return a copy that outlives the handler
     */
    public byte[] toByteArray() {
        return pixels.toArray(ValueLayout.JAVA_BYTE);
    }

    /**
     * @return the sender's frame id, or 0 when the frame carried no metadata trailer
     */
    public long frameId() {
        return frameId;
    }

    /**
     * When the sender says it captured this frame, in microseconds on the sender's own clock.
     *
     * <p>Differences between stamps from one sender are what this supports. It is not comparable
     * with a local clock, and it is 0 when the frame carried no metadata.
     *
     * @return the sender's capture time
     */
    public long timestampUs() {
        return timestampUs;
    }

    /**
     * The tag the sender attached to this frame.
     *
     * <p>Already copied, so unlike {@link #pixels()} it outlives the handler. A tag is dropped
     * unless the far end declared that it reads tags.
     *
     * @return the tag, or empty when the frame carried none
     */
    public Optional<byte[]> userData() {
        return Optional.ofNullable(userData).map(bytes -> bytes.clone());
    }
}

File: sdks/android/reactor-sdk-android/src/main/kotlin/inc/reactor/sdk/android/VideoFrame.kt
package inc.reactor.sdk.android

import java.nio.ByteBuffer

/**
 * One decoded video frame, in BGRA.
 *
 * **[pixels] is borrowed.** It is a direct [ByteBuffer] over memory the native layer owns, valid
 * only until the handler returns — the FFI frees it the moment it does. Keeping the buffer and
 * reading it later is a use-after-free, which reproduces under load and not in tests. Copy what
 * you keep, or consume it in place with the renderer helpers in `reactor-sdk-android-media`.
 *
 * The trailer fields are 0 when the sender attached no metadata. A tag is dropped unless the far
 * end declared that it reads tags, so an absent [frameId] is a statement about the sender rather
 * than about this frame.
 */
public class VideoFrame internal constructor(
    /** The track this arrived on. Every recvonly video track decodes into one callback. */
    public val trackName: String,
    /** Borrowed BGRA pixels: `width * height * 4` bytes. Valid only for this call. */
    public val pixels: ByteBuffer,
    public val width: Int,
    public val height: Int,
    /** The sender's frame counter, or 0 when it attached no metadata. */
    public val frameId: Long,
    /**
     * When the sender says it captured this, in microseconds **on the sender's own clock**.
     *
     * Differences between stamps from one sender are what it supports. It is not comparable with
     * a local clock, and subtracting it from `System.currentTimeMillis()` is meaningless.
     */
    public val timestampUs: Long,
    /** The sender's opaque tag, or null when there was none. Borrowed, like [pixels]. */
    public val userData: ByteArray?,
) {
    /** A copy of [pixels] that outlives the callback. */
    public fun copyPixels(): ByteArray {
        val duplicate = pixels.duplicate()
        return ByteArray(duplicate.remaining()).also { duplicate.get(it) }
    }

    override fun toString(): String = "VideoFrame($trackName, ${width}x$height, frameId=$frameId, timestampUs=$timestampUs)"
}
